Modify DAS_Tool input structure, bump version, switch to nf-test

prototaxites opened the pull request:
1. Old version of dastool/dastool module took the proteins fasta on a separate input channel, but this has to match the required contigs input so I've moved the protein input to the same channel.
2. Bump to latest DAS_Tool version
3. Relabel two mislabelled output channels in dastool/dastool
4. Change protein fasta decompression to check if input proteins are compressed in dastool/dastool
5. Switch old tests to new nf-tests for all dastool modules

## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

prototaxites commented:
Not sure these CI tests are working correctly - they run fine on Gitpod, but here it doesn't seem to generate output files?

SPPearce commented:
> Not sure these CI tests are working correctly - they run fine on Gitpod, but here it doesn't seem to generate output files?

Conda seems to have errored entirely, with lots of complaints about missing libraries before it tries.
Docker and singularity seem to have a difference in:
`"test_proteins.faa:md5,542ed4ddbf454feed5ef279361fdd1d5"`

prototaxites commented:
Docker and Conda tests seem to be passing now. Singularity fails as the "test_proteins.faa" snapshots differ - must be something different in the singularity environment compared to the docker/conda one?

```
! 35             "test_proteins.faa:md5 ! 35             "test_proteins.faa:md5
,542ed4ddbf454feed5ef279361fdd1d5"      ,b23fa2a776b252204d5a2adc8355c98b"     
```

prototaxites commented:
Replaced the md5 test with a "contains" test, as the files have the same contents between the two run types, but in different orders for some reason.

famosab reviewed (commented):
Looks good already. I was wondering if you could also add a stub section to the modules and a stub test to the tests :)

prototaxites commented on modules/nf-core/dastool/dastool/main.nf:
```diff
@@ -4,12 +4,11 @@ process DASTOOL_DASTOOL {
     input:
-    tuple val(meta), path(contigs), path(bins)
-    path(proteins)
+    tuple val(meta), path(contigs), path(proteins), path(bins)
```
In my mind the contigs and proteins are linked (proteins are Prodigal predictions from the contigs) so they looked better together... appreciate they're optional though so perhaps putting the optional argument at the end is cleaner. Happy to change if you think that's better!

famosab replied:
I would append it - guess that is a matter of preference but if I write my pipeline then I am happy to just have empty [] in the end for optional stuff that I don't need :)

prototaxites commented on modules/nf-core/dastool/dastool/main.nf:
```diff
@@ -19,8 +18,8 @@ process DASTOOL_DASTOOL {
-    tuple val(meta), path("*.candidates.faa")           , optional: true, emit: fasta_proteins
-    tuple val(meta), path("*.faa")                      , optional: true, emit: candidates_faa
+    tuple val(meta), path("*.candidates.faa")           , optional: true, emit: candidates_faa
+    tuple val(meta), path("*proteins.faa")              , optional: true, emit: fasta_proteins
```
file is called "${prefix}_proteins.faa" - no .

famosab replied:
ah good to know - changed it to be consistent. then i would suggest to add the `_` to the output

prototaxites commented on modules/nf-core/dastool/scaffolds2bin/tests/main.nf.test:
This is the old name for the subtool that's only available in older versions of dastool. Honestly not sure why there's a module - don't think it's used in any pipelines.

famosab replied:
maybe we can deprecate it then?

famosab replied:
https://nf-co.re/docs/guidelines/components/module_deprecation

prototaxites replied:
paging @jfy133 - do we need this module any more?

jfy133 replied:
No, can be deprecated!

I think they actually renamed the script: 
![image](https://github.com/user-attachments/assets/918ac133-8763-4eb5-b50f-82a08780d27a)

But even so it should use the new name not the old one - so this one can be deprecated.

prototaxites commented on modules/nf-core/dastool/dastool/tests/main.nf.test:
```diff
@@ -0,0 +1,136 @@
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    process.out.summary,
+                    process.out.contig2bin,
+                    process.out.eval,
+                    process.out.bins,
+                    process.out.pdfs,
+                    process.out.fasta_proteins,
+                    process.out.candidates_faa,
+                    process.out.fasta_archaea_scg,
+                    process.out.fasta_bacteria_scg,
+                    process.out.b6,
+                    process.out.versions
```
Perhaps I'm missing something, but all of these are emitted by the process? Some of them (evals & bins) are only emitted when certain arguments are supplied, hence the lack of hashes in the snapshot.

famosab replied:
ok makes sense, but then if all of them are asserted we can also simply do
```
{ assert snapshot( process.out ).match() }
```

prototaxites replied:
Ah, it's because I left out `process.out.log` as it contains a timestamp, so can't match the whole process.out - have added a "contains" test now to check the contents.

famosab replied:
then its fine asserting the snapshots is still better then, and you can assert the log aswell with contains

prototaxites commented:
OK - added stubs and tests for them (this required adding a stub to metabat2/metabat2, which the test depends on). Added a contains test for the dastool/dastool log file. Moved the proteins input channel to the end.

famosab commented on modules/nf-core/dastool/dastool/tests/nextflow.config:
Small request here: https://nf-co.re/docs/guidelines/components/modules#configuration-of-extargs-in-tests :)

famosab commented on modules/nf-core/metabat2/metabat2/main.nf:
maybe we can open an issue that the stub test for this module also needs to be created - but I think you dont have to do that here :)

famosab commented:
I am not sure why you need the stub section in the other module - yes it would be better but you should be able to use the Process outputs without stubs from your setup block (or maybe I am mistaken there).

prototaxites commented:
Seems that when you make a stub run with `options = "-stub"` all modules run in stub mode, including those in the nf-test setup - so the missing stub in METABAT2_METABAT2 causes it to fail.

prototaxites commented:
pytest for metabat2/metabat2 is failing - looks like it's not finding micromamba so the test dies.

famosab commented:
There are also some other tests failing.

I'll check how we can fix the metabat test.

famosab commented:
Maybe we can try adding an environment file to the metabat module. Its very old (last edit was in 2021 if I see that correctly) so it uses a rather old structure.

prototaxites commented:
Both the Singularity and Docker pytests for metabat2 do pass - it's a conda error specifically: https://github.com/nf-core/modules/actions/runs/12256931253/job/34193327009?pr=7163

I wonder if we can just ignore this one test for now, given that adding the stub shouldn't affect the metabat2 tests? And leave the issue with the runners lacking micromamba to be fixed in a separate PR?

famosab commented:
In theory i am with you but in practice we would need to bypass branch protection rules to do that in this case. I think we have to ask in the slack how to handle this special case :/

jfy133 commented:
I am working on mag now waiting for tests, so I'll see if I can update METABAT_METABAT in a separate PR for you :)

jfy133 commented:
MB2: https://github.com/nf-core/modules/pull/7196

jfy133 commented:
OK the METABAT2 PR is in the merge queue now, thanks for the fast review @prototaxites !

I guess in 10 minutes you can update this PR and hopefully you'll get your nice green tick :)

prototaxites commented:
All done!

famosab commented on modules/nf-core/dastool/scaffolds2bin/tests/main.nf.test:
```diff
@@ -0,0 +1,59 @@
+
+        then {
+            assertAll(
+                { assert process.failed },
```
why did you decide to do that assertion?

famosab reviewed (commented):
Looks good, its just the one test that I am unsure about?

famosab reviewed (approved):
Nevermind just realized that is the deprecated tool :) LGTM!
